src/CustomFunction.ts

```diff
@@ -0,0 +1,211 @@
+
+      return client.functions.completeError({
+        token,
+        error,
+        function_execution_id: functionExecutionId,
+      });
+    };
+  }
+}
+
+/** Helper Functions */
+export function validate(callbackId: string, middleware: CustomFunctionExecuteMiddleware): void {
+  // Ensure callbackId is valid
+  if (typeof callbackId !== 'string') {
+    const errorMsg = 'CustomFunction expects a callback_id as the first argument';
+    throw new CustomFunctionInitializationError(errorMsg);
+  }
+
+  // Ensure middleware argument is either a function or an array
+  if (typeof middleware !== 'function' && !Array.isArray(middleware)) {
+    const errorMsg = 'CustomFunction expects a function or array of functions as the second argument';
+    throw new CustomFunctionInitializationError(errorMsg);
+  }
+
+  // Ensure array includes only functions
+  if (Array.isArray(middleware)) {
+    middleware.forEach((fn) => {
+      if (!(fn instanceof Function)) {
+        const errorMsg = 'All CustomFunction middleware must be functions';
+        throw new CustomFunctionInitializationError(errorMsg);
+      }
+    });
+  }
+}
+
+/**
+ * `processFunctionMiddleware()` invokes each callback for lifecycle event
+ * @param args workflow_step_edit action
+ */
+export async function processFunctionMiddleware(
+  args: AllCustomFunctionMiddlewareArgs,
+  middleware: CustomFunctionMiddleware,
+): Promise<void> {
+  const { context, client, logger } = args;
+  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
+  const lastCallback = callbacks.pop();
+
+  if (lastCallback !== undefined) {
+    await processMiddleware(
+      callbacks, args, context, client, logger,
+      async () => lastCallback({ ...args, context, client, logger }),
+    );
+  }
+}
+
+export function isFunctionEvent(args: AnyMiddlewareArgs): args is AllCustomFunctionMiddlewareArgs {
+  return VALID_PAYLOAD_TYPES.has(args.payload.type);
+}
+
+function selectToken(context: Context): string | undefined {
+  // If attachFunctionToken = false, fallback to botToken or userToken
+  return context.functionBotAccessToken ? context.functionBotAccessToken : context.botToken || context.userToken;
+}
+
+/**
+ * `enrichFunctionArgs()` takes in a function's args and:
+ *  1. removes the next() passed in from App-level middleware processing
+ *    - events will *not* continue down global middleware chain to subsequent listeners
+ *  2. augments args with step lifecycle-specific properties/utilities
+ * */
+export function enrichFunctionArgs(args: any): AllCustomFunctionMiddlewareArgs {
+  const { next: _next, ...functionArgs } = args;
+  const enrichedArgs: any = { ...functionArgs };
+  const token = selectToken(functionArgs.context);
+
+  // Making calls with a functionBotAccessToken establishes continuity between
+  // a function_executed event and subsequent interactive events (actions)
+  const client = new WebClient(token, { ...functionArgs.client });
```

Comment:
This is the first time to see this way to initialize a new instance. Other code in bolt-js uses clientOptions for the second argument instead. I haven't verified on my end yet but have you confirmed this code can make the same effect with `new WebClient(token, clientOptions)`?

Reply:
I'm 99% sure I tested this when you brought it up as a concern back in January, but we will make doubly-sure before we cut the stable release.

Reply:
Going to take a slightly different tact here: expose the web client options as assembled in the bolt App constructor via a getter method, and simply re-use those here instead.